### PR TITLE
Showing tables in comparison by demand

### PR DESCRIPTION
**Pull request by detininroman**
By default tables are hidden. User can choose tables to be shown or show all.

### CODE REVIEW

File: perftracker/templates/comparison_id.html [hunk 1]
```diff
@@ -57,16 +61,76 @@ <h3 class='pt_collapse expanded' data-toggle='collapse' data-target='#data_{{ g.
...
+
+function hide_all_tables() {
+    shown_tables.map(table_id => hide_table(table_id));
+}
+
+function show_all_tables() {
+    $.getJSON("/api/v{{ api_ver }}/{{ project.id }}/comparison/{{ obj.id }}/group/{0}/section/{1}".ptFormat(-1, -1),
+        function (data) {
+            data = $.parseJSON(data);
+            for (const [table_id, _] of Object.entries(data)) {
```
Comment (istillc0de) [inline perftracker/templates/comparison_id.html]: "for .. of" is ES6 feature. It may be above the original compatibility goal of the PT project (for example ES5 was a requred JS version)  - need to doublecheck  that with the author

File: perftracker/templates/comparison_id.html [hunk 2]
```diff
@@ -57,16 +61,76 @@ <h3 class='pt_collapse expanded' data-toggle='collapse' data-target='#data_{{ g.
...
+                show_table(table_id, data);
+            }
+        }
+    );
+}
+
+function toggle_table(group_id, section_id)
+{
+    let table_id = 'tests_{0}_{1}'.ptFormat(group_id, section_id);
```
Comment (istillc0de) [inline perftracker/templates/comparison_id.html]: 'let' is ES6 feature

File: perftracker/templates/comparison_id.html (hunk 1 above)
Comment (istillc0de) [inline perftracker/templates/comparison_id.html]: Use Object.keys() instead of .entries()

File: perftracker/templates/comparison_id.html [hunk 3]
```diff
@@ -89,20 +153,16 @@ <h3 class='pt_collapse expanded' data-toggle='collapse' data-target='#data_{{ g.
...
-        s += '<td>' + func(ar[n]) + '</td>';
-    return s + '</tr>';
+    pt_configure_table_async(
+        '#{0}'.ptFormat(table_id),
+        data[table_id]['pageable'],
+        "/api/v{{ api_ver }}/{{ project.id }}/comparison/{{ obj.id }}/group/0/test/",
+        data[table_id]['table_data'],
+    );
+
+    shown_tables.push(table_id);
```
Comment (istillc0de) [inline perftracker/templates/comparison_id.html]: to add/check for existance we don't usually use array, but set.

File: perftracker/templates/comparison_id.html [hunk 4]
```diff
@@ -155,24 +215,24 @@ <h3 class='pt_collapse expanded' data-toggle='collapse' data-target='#data_{{ g.
...
   pt_configure_chart_async('chart_{{ g.id }}_{{ s.id }}', {{ s.chart_type }}, {% if s.has_failures %}true{% else %}false{% endif %},
      {{ s.x_axis_categories|safe }}, '{{ s.x_axis_name }}', '{{ s.x_axis_type }}', {{ s.x_axis_rotate }}, '{{ s.y_axis_name }}', [
 {% for serie in s.series %}      { name: '{{ serie.legend|safe }}', data: {{ serie.data|safe }}{% if s.chart_trend_line %}, trend: true{% endif %}},
 {% endfor %}     ]);
-{% endif %}{% if s.table_type == PTCmpTableType.SHOW %}
-  pt_configure_table_async('#tests_{{ g.id }}_{{ s.id }}', {% if s.pageable %}1{% else %}0{% endif %}, testlink, [
-{% for t_tag, t in s.tests.items %}      {{ t.table_data|safe }},
-{% endfor %}    ]);
+{% endif %}
+{% if s.table_type == PTCmpTableType.SHOW %}
```
Comment (istillc0de) [inline perftracker/templates/comparison_id.html]: seems like wrong location - inside 2 loops, while it must be once per ready()

File: perftracker/templates/comparison_id.html [hunk 5]
```diff
@@ -57,16 +61,76 @@ <h3 class='pt_collapse expanded' data-toggle='collapse' data-target='#data_{{ g.
...
+}
+
+function toggle_table(group_id, section_id)
+{
+    let table_id = 'tests_{0}_{1}'.ptFormat(group_id, section_id);
+
+    if (shown_tables.includes(table_id)) {
+        hide_table(table_id);
+    } else {
+        $.getJSON("/api/v{{ api_ver }}/{{ project.id }}/comparison/{{ obj.id }}/group/{0}/section/{1}".ptFormat(group_id, section_id),
```
Comment (istillc0de) [inline perftracker/templates/comparison_id.html]: how about error handling? 
For example: add a div with 'unable to get table data...'

File: perftracker/templates/comparison_id.html [hunk 6]
```diff
@@ -57,16 +61,76 @@ <h3 class='pt_collapse expanded' data-toggle='collapse' data-target='#data_{{ g.
...
-      <div class='container collapse in pt_div_cmp' id='results_{{ g.id }}_{{ s.id }}'>
-        <table id="tests_{{ g.id }}_{{ s.id }}" class="display dataTable" cellspacing="0" width="100%">
+    <div class='container collapse in pt_div_cmp' id='results_{{ g.id }}_{{ s.id }}'></div>
+    {% endspaceless %}
+  {% endfor %}
+  </div>
+{% endfor %}
+
+<script>
+function pt_test_details_draw_row(title, ar, func)
```
Comment (istillc0de) [inline perftracker/templates/comparison_id.html]: can be moved to perftracker.js, along with pt_test_details_draw

File: perftracker/templates/comparison_id.html [hunk 7]
```diff
@@ -57,16 +61,76 @@ <h3 class='pt_collapse expanded' data-toggle='collapse' data-target='#data_{{ g.
...
+            }
+        );
+    }
+}
+
+function hide_table(table_id) {
+    if (!shown_tables.includes(table_id)) {
+        return;
+    }
+    $('#{0}'.ptFormat(table_id)).remove();
```
Comment (istillc0de) [inline perftracker/templates/comparison_id.html]: why remove element which is already rendered? just hide it.
when shown again, check if this element is present and hidden, don't go server for data then, just show it.

File: perftracker/views.py [hunk 8]
```diff
@@ -130,6 +130,38 @@ def pt_comparison_all_html(request, project_id):
...
+        obj = ComparisonModel.objects.get(pk=cmp_id)
+    except ComparisonModel.DoesNotExist:
+        raise Http404
+
+    ret = {}
+    cmp_view = PTComparisonServSideView(obj)
+
+    # data for all the tables
+    if int(group_id) == -1 and int(section_id) == -1:
+        for _, group in cmp_view.groups.items():
```
Comment (istillc0de) [inline perftracker/views.py]: use .values() instead

File: perftracker/views.py [hunk 9]
```diff
@@ -130,6 +130,38 @@ def pt_comparison_all_html(request, project_id):
...
+    try:
+        obj = ComparisonModel.objects.get(pk=cmp_id)
+    except ComparisonModel.DoesNotExist:
+        raise Http404
+
+    ret = {}
+    cmp_view = PTComparisonServSideView(obj)
+
+    # data for all the tables
+    if int(group_id) == -1 and int(section_id) == -1:
```
Comment (istillc0de) [inline perftracker/views.py]: this is not fitting well into REST concept. consider separate URL for that instead of -1/-1.
Actual handler still can be the same for both URLs

File: perftracker/views.py [hunk 10]
```diff
@@ -130,6 +130,38 @@ def pt_comparison_all_html(request, project_id):
...
+                ret[f'tests_{group.id}_{section.id}'] = {
+                    'table_data': [t.table_data for _, t in tests],
+                    'pageable': int(section.pageable),
+                }
+
+    # data for the particular table
+    else:
+        _, group = list(cmp_view.groups.items())[int(group_id)]
+        _, section = list(group.sections.items())[int(section_id)]
+        tests = section.tests.items()
```
Comment (istillc0de) [inline perftracker/views.py]: .values()

File: perftracker/views.py [hunk 11]
```diff
@@ -130,6 +130,38 @@ def pt_comparison_all_html(request, project_id):
...
 
 
+def pt_comparison_tables_info_json(request, api_ver, project_id, cmp_id, group_id, section_id):
+    try:
+        obj = ComparisonModel.objects.get(pk=cmp_id)
+    except ComparisonModel.DoesNotExist:
+        raise Http404
+
+    ret = {}
+    cmp_view = PTComparisonServSideView(obj)
```
Comment (istillc0de) [inline perftracker/views.py:140]: This is HUGE database and Python work to get all details for all jobs in comparison.
Using it to get individual table data seems like an overkill.

File: perftracker/views.py [hunk 12]
```diff
@@ -130,6 +130,38 @@ def pt_comparison_all_html(request, project_id):
...
+            for _, section in group.sections.items():
+                tests = section.tests.items()
+                ret[f'tests_{group.id}_{section.id}'] = {
+                    'table_data': [t.table_data for _, t in tests],
+                    'pageable': int(section.pageable),
+                }
+
+    # data for the particular table
+    else:
+        _, group = list(cmp_view.groups.items())[int(group_id)]
```
Comment (istillc0de) [inline perftracker/views.py]: are we sure items() are sequentially-ordered? I don't think so, and we cannot use index-based access.

File: perftracker/templates/comparison_id.html [hunk 13]
```diff
@@ -57,16 +61,76 @@ <h3 class='pt_collapse expanded' data-toggle='collapse' data-target='#data_{{ g.
...
+    shown_tables = shown_tables.filter(item => item !== table_id);
+}
+
+function show_table(table_id, data)
```
Comment (istillc0de) [inline perftracker/templates/comparison_id.html]: consider to change the format of 'table_id' (and rename it to smth like section_full_id). It is currently 'table_2_5', but better be '2_5'. In that case you can just concatenate it with any prefix (e.g. #table_ or #result_), no need for ptFormat() or match().

File: perftracker/templates/comparison_id.html [hunk 14]
```diff
@@ -57,16 +61,76 @@ <h3 class='pt_collapse expanded' data-toggle='collapse' data-target='#data_{{ g.
...
+        s += '<td>' + func(ar[n]) + '</td>';
+    return s + '</tr>';
+}
+
+function hide_all_tables() {
+    shown_tables.map(table_id => hide_table(table_id));
+}
+
+function show_all_tables() {
+    $.getJSON("/api/v{{ api_ver }}/{{ project.id }}/comparison/{{ obj.id }}/group/{0}/section/{1}".ptFormat(-1, -1),
```
Comment (istillc0de) [inline perftracker/templates/comparison_id.html]: error/timeout handling?

File: perftracker/views.py [hunk 15]
```diff
@@ -130,6 +130,38 @@ def pt_comparison_all_html(request, project_id):
...
+            for section in group.sections.values():
+                tests = section.tests.values()
+                ret[f'{group.id}_{section.id}'] = {
+                    'table_data': [t.table_data for t in tests],
+                    'pageable': int(section.pageable),
+                }
+
+    # data for the particular table
+    else:
+        [group] = filter(lambda g: g.id == int(group_id), cmp_view.groups.values())
```
Comment (istillc0de) [inline perftracker/views.py]: now it is correct code but can be merged with previous section, as filter()s are basically loops with if.

File: perftracker/templates/comparison_id.html [hunk 16]
```diff
@@ -89,60 +164,17 @@ <h3 class='pt_collapse expanded' data-toggle='collapse' data-target='#data_{{ g.
             </tr>
           </thead>
         </table>
-      </div>
-    {% endif %}
-    {% endspaceless %}
-  {% endfor %}
-  </div>
-{% endfor %}
+        <span style="display:block; height: 40px;" id=space_{0}></span>
```
Comment (istillc0de) [inline perftracker/templates/comparison_id.html]: I think we can nest padding item and table and avoid showing/hiding 2 elements instead of 1.
Or table style can be provide sufficient padding instead of this span

File: perftracker/templates/comparison_id.html [hunk 17]
```diff
@@ -57,16 +61,87 @@ <h3 class='pt_collapse expanded' data-toggle='collapse' data-target='#data_{{ g.
...
+            show_table(test_case_id, None);
+            return;
+        }
+        $.getJSON("/api/v{{ api_ver }}/{{ project.id }}/comparison/{{ obj.id }}/group/{0}/section/{1}".ptFormat(group_id, section_id),
+            function (data) {
+                data = $.parseJSON(data);
+                show_table(test_case_id, data[test_case_id]);
+            }
+        ).error(function (xhr) {
+            console.log('Error: ' + xhr.status + ' ' + xhr.responseText)
```
Comment (istillc0de) [inline perftracker/templates/comparison_id.html]: this is not a 'real' error handling. See pt_configure_table for better one.

File: perftracker/templates/comparison_id.html [hunk 18]
```diff
@@ -155,26 +187,27 @@ <h3 class='pt_collapse expanded' data-toggle='collapse' data-target='#data_{{ g.
     table.ajax.url(link).load(null, true);
 }
 
+var shown_tables = new Set();
```
Comment (istillc0de) [inline perftracker/templates/comparison_id.html]: I am really uncomfortable that you describe the state with 3 sets.

shown/hidden can be obtained via
$(element).is(":hidden");   or similar.

I would generate HTML with single Object like 
`
`
```suggestion
var data = {
...
  '1_21': { 
       chart: [....],   // generated by template
       table: null   // <-- this one will be requested later by XHR
       // if needed (may be not, we can determine visibility from the control itself)
       chart_shown: true,
       table_shown: false
    },
   ...
}

```

File: perftracker/static/js/perftracker.js [hunk 19]
```diff
@@ -800,4 +800,63 @@ function update_nav_bar(auth_response) {
...
+
+    s += "</div>"; /* row */
+
+    s += "</div>"; /* pt_slider */
+
+    return s;
+}
+
+function pt_table_loading_error_handler(test_case_id, error) {
+    if ($("#error_{0}".ptFormat(test_case_id)).length) {
```
Comment (istillc0de) [inline perftracker/static/js/perftracker.js]: error_x  can become non-unique accross page (imagine charts can fail too)
better to name table_error_x

File: perftracker/views.py [hunk 20]
```diff
@@ -130,6 +130,36 @@ def pt_comparison_all_html(request, project_id):
...
+    try:
+        obj = ComparisonModel.objects.get(pk=cmp_id)
+    except ComparisonModel.DoesNotExist:
+        raise Http404
+
+    ret = {}
+    cmp_view = PTComparisonServSideView(obj)
+
+    all_tables_search = group_id is None and section_id is None
+    single_table_search = not all_tables_search
```
Comment (istillc0de) [inline perftracker/views.py]: useless, please remove. optimization is not worth the trouble

File: perftracker/templates/comparison_id.html [hunk 21]
```diff
@@ -57,16 +61,105 @@ <h3 class='pt_collapse expanded' data-toggle='collapse' data-target='#data_{{ g.
...
+    const tables_are_rendered = Object.keys(data).length === {{ test_cases_amount }};
+    var request_failed = false;
+
+    if (!tables_are_rendered) {
+        $.ajax({
+            url: "/api/v{{ api_ver }}/{{ project.id }}/comparison/{{ obj.id }}/tables",
+            data: null,
+            type: 'GET',
+            async: false,
+            timeout: 2000,
```
Comment (istillc0de) [inline perftracker/templates/comparison_id.html]: lets do 10000 - it can return a lot of data

File: perftracker/templates/comparison_id.html [hunk 22]
```diff
@@ -57,16 +61,105 @@ <h3 class='pt_collapse expanded' data-toggle='collapse' data-target='#data_{{ g.
...
+            success: function (response) {
+                response = $.parseJSON(response);
+                Object.keys(response).map(test_case_id => {
+                    data[test_case_id] = response[test_case_id];
+                });
+            },
+            error: function (data, status, error) {
+                {% for g_tag, g in cmp_view.groups.items %} {% for s_tag, s in g.sections.items %}
```
Comment (istillc0de) [inline perftracker/templates/comparison_id.html]: I did not expect you still use code generation. 
The only generation we need is 
var data = {....} 
thing

File: perftracker/templates/comparison_id.html [hunk 23]
```diff
@@ -155,26 +203,27 @@ <h3 class='pt_collapse expanded' data-toggle='collapse' data-target='#data_{{ g.
...
+function show_all_charts() {
+    {% for g_tag, g in cmp_view.groups.items %} {% for s_tag, s in g.sections.items %}
+      pt_configure_chart_async('chart_{{ g.id }}_{{ s.id }}', {{ s.chart_type }}, {% if s.has_failures %}true{% else %}false{% endif %},
+        {{ s.x_axis_categories|safe }}, '{{ s.x_axis_name }}', '{{ s.x_axis_type }}', {{ s.x_axis_rotate }}, '{{ s.y_axis_name }}', [
+        {% for serie in s.series %}      { name: '{{ serie.legend|safe }}', data: {{ serie.data|safe }}{% if s.chart_trend_line %},
+        trend: true{% endif %} }, {% endfor %} ]);
+    {% endfor %} {% endfor %}
+}
+
+var data = {};
```
Comment (istillc0de) [inline perftracker/templates/comparison_id.html]: pls rename more descriptively

**Review by istillc0de — APPROVED** [review]
